common/ledger/blkstorage/cache.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package blkstorage
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+)
+
+const (
+	estimatedBlockSize = 512 * 1024
+)
+
+type Cache struct {
+	cacheLock    sync.RWMutex
+	disabled     bool
+	cache        map[uint64]cachedBlock
+	sizeBytes    int
+	maxSeq       uint64
+	maxSizeBytes int
+}
+
+type cachedBlock struct {
+	block     *common.Block
+	blockSize int
+}
+
+func NewCache(maxSizeBytes int) *Cache {
+	isCacheDisabled := maxSizeBytes == 0
+
+	return &Cache{
+		disabled:     isCacheDisabled,
+		cache:        make(map[uint64]cachedBlock, maxSizeBytes/estimatedBlockSize),
+		maxSizeBytes: maxSizeBytes,
+	}
+}
+
+func (c *Cache) retrieveCachedBlock(seq uint64) (*common.Block, bool) {
+	if c.disabled {
+		return nil, false
+	}
+
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+
+	cachedBlock, exists := c.cache[seq]
+	return cachedBlock.block, exists
+}
+
+func (c *Cache) cacheBlock(block *common.Block, blockSize int) {
+	if c.disabled {
+		return
+	}
+
+	seq := block.Header.Number
+
+	if c.maxSeq > seq {
+		return
+	}
+
+	// Insert the block to the cache
+	c.maxSeq = seq
```

Comment:
If it's higher it means we have appended a block out of order. If we did that, we should panic here, no?

Reply:
On second thought, we can't do this because the block cache is not initialized with the last block number. I prefer to keep it this way for simplicity.